openslides_backend/services/datastore/extended_adapter.py

```diff
@@ -0,0 +1,476 @@
+import builtins
+import re
+from collections import defaultdict
+from copy import deepcopy
+from typing import Any, Dict, List, Literal, Optional, Tuple
+
+from datastore.shared.postgresql_backend import SqlQueryHelper
+from datastore.shared.util import DeletedModelsBehaviour
+
+from ...shared.exceptions import DatastoreException
+from ...shared.filters import And, Filter, FilterOperator
+from ...shared.interfaces.logging import LoggingModule
+from ...shared.patterns import Collection, FullQualifiedId
+from ...shared.typing import DeletedModel, ModelMap
+from .adapter import DatastoreAdapter
+from .commands import GetManyRequest
+from .handle_datastore_errors import raise_datastore_error
+from .interface import Engine, LockResult, PartialModel
+
+MODEL_FIELD_SQL = "data->>%s"
+COMPARISON_VALUE_SQL = "%s::text"
+
+
+MappedFieldsPerFqid = Dict[FullQualifiedId, List[str]]
+
+
+class ExtendedDatastoreAdapter(DatastoreAdapter):
+    """
+    Subclass of the datastore adapter to extend the functions with the usage of the changed_models.
+
+    Restrictions:
+    -   get_deleted_models only works one way with the changed_models: if the model was not deleted
+        in the datastore, but is deleted in the changed_models. The other way around does not work
+        since a deleted model in the changed_models is marked via DeletedModel() and does not store
+        any data.
+    -   all filter-based requests may take two calls to the datastore to succeed. The first call is
+        always necessary, since the changed_models are never complete. If, however, a model in the
+        changed_models matches the filter which it did not in the database AND some fields are
+        missing in the changed_models which are needed through the mapped_fields, a second request
+        is needed to fetch the missing fields. This can be circumvented by always storing (more or
+        less) "full" models in the changed_data, meaning all relevant fields which are requested in
+        future calls are present. This is the case for most applications in the backend.
+    -   filters are only evaluated separately on the changed_models and the datastore. If, for
+        example, a model in the datastore does not fit the filter, but through a change in the
+        changed_models would fit it, BUT does not fit the filter from the changed_models alone, it
```

Comment:
For me it seems to be more consistent, if we would apply 

- all fields from datastore content, that are NOT in corresponding  changed_models respecting None-values and
- the missing fields including the filter fields for all changed_models, that are not in datastore content

 to the changed_models. Then we could filter the changed_models and the results are reflecting the updated data.
The data to read seems nearly identical.

Example with filtered ids as set:
datastore content: {1, 2, 3}
changed_models: {2, 3, 4, 5}
filtered_changed_models: {2, 4}
FInal result set with ids: (datastore_content - changed_models) | filtered_changed_models = {1, 2, 4}
This is the expected result for a filter on current data, i.e. with use_changed_models = True.

Reply:
This will not work, if I understand you correctly, since you do not have all data and do not know what you need to fetch. Imagine the following:

datastore: `{"model/1": {"a": 1}}`
changed_models: `{"model/1": {"b": 2}}`
filter request: `a=1 AND b=2`

The filter request to the datastore will return an empty result, as well as the one to the changed_models, but `model/1` would have to be returned for a correct result. But as written further down in the comment, this should not be a problem for us since we only filter on static/readonly fields like `meeting_id` (at least I never encountered any other filter yet).